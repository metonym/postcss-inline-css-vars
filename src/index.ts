import type { Plugin } from "postcss";

const isCssVar = (value: string) => value.includes("var(");

const ROOT_SELECTOR = ":root";

const RE_CSS_VARS = /var\((--[^)]+)\)/g;

const extractVarNames = (value: string): string[] => {
  const matches = [...value.matchAll(RE_CSS_VARS)];
  return matches.map((match) => match[1]);
};

export const inlineCssVars = (): Plugin => {
  return {
    postcssPlugin: "postcss-inline-css-vars",
    Once(root) {
      /**
       * Check if there are any :root rules with additional selectors.
       * In this case, we don't want to process the :root rules since
       * we cannot safely inline the CSS variables.
       * @example
       * :root { --primary-color: #000000; }
       * :root.dark { --primary-color: #ffffff; }
       */
      let has_complex_root_rules = false;

      root.walkRules((rule) => {
        if (
          rule.selector.includes(ROOT_SELECTOR) &&
          rule.selector !== ROOT_SELECTOR
        ) {
          has_complex_root_rules = true;
          return false;
        }
      });

      if (has_complex_root_rules) return;

      let css_vars: Record<string, string> = {};
      let has_root_vars = false;

      root.walkRules(ROOT_SELECTOR, (rule) => {
        has_root_vars = true;
        rule.walkDecls((decl) => {
          css_vars[decl.prop] = decl.value;
        });
        rule.remove();
      });

      if (!has_root_vars) return;

      let changed = true;

      while (changed) {
        changed = false;

        for (const [prop, value] of Object.entries(css_vars)) {
          if (isCssVar(value)) {
            const new_value = value.replace(
              RE_CSS_VARS,
              (match, varName) => {
                return css_vars[varName] || match;
              }
            );

            if (new_value !== value) {
              css_vars[prop] = new_value;
              changed = true;
            }
          }
        }
      }

      root.walkDecls((decl) => {
        if (isCssVar(decl.value)) {
          const var_names = extractVarNames(decl.value);
          const has_undefined_var = var_names.some(
            (var_name) => !css_vars[var_name]
          );

          if (has_undefined_var) {
            // Remove declaration if the CSS variable does not exist.
            decl.remove();
          } else {
            // Replace all variables with their resolved values.
            decl.value = decl.value.replace(RE_CSS_VARS, (match, var_name) => {
              return css_vars[var_name] || match;
            });
          }
        }
      });
    },
  };
};
