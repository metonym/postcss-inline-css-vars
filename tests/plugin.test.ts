import postcss from "postcss";
import { inlineCssVars } from "../src";

describe("inlineCssVars", () => {
  const process = (css: string) => {
    return postcss(inlineCssVars()).process(css.trim()).css.trim();
  };

  test("removes :root rule after processing", () => {
    const source = `
      :root { --primary-color: #ff0000; }
      h1 { color: red; }
    `;

    expect(process(source)).toBe("h1 { color: red; }");
  });

  test("replaces CSS variable with its value", () => {
    const source = `
      :root { --primary-color: #ff0000; }
      h1 { color: var(--primary-color); }
    `;

    expect(process(source)).toBe("h1 { color: #ff0000; }");
  });

  test("handles multiple CSS variables", () => {
    const source = `
      :root {
        --primary-color: #ff0000;
        --secondary-color: #00ff00;
      }
      h1 { color: var(--primary-color); }
      p { color: var(--secondary-color); }
    `;

    expect(process(source)).toMatchSnapshot();
  });

  test("removes declarations with undefined variables", () => {
    const source = `
      :root { 
        --defined-color: #ff0000;
      }
      .button { 
        color: var(--undefined-color);
        background: var(--defined-color);
      }
    `;

    expect(process(source)).toMatchSnapshot();
  });

  test("removes declarations with any undefined variable in multi-var declarations", () => {
    const source = `
      :root { 
        --spacing-x: 10px;
      }
      .box { 
        margin: var(--spacing-y) var(--spacing-x);
        padding: var(--spacing-x) var(--spacing-x);
      }
    `;

    expect(process(source)).toMatchSnapshot();
  });

  test("removes declarations with undefined variables instead of preserving them", () => {
    const source = `
      :root { --primary-color: #ff0000; }
      h1 { 
        color: var(--undefined-color);
        background: var(--primary-color);
      }
    `;

    expect(process(source)).toMatchSnapshot();
  });

  test("handles nested CSS variables", () => {
    const source = `
      :root {
        --primary: #ff0000;
        --button-color: var(--primary);
      }
      button { color: var(--button-color); }
    `;
    const result = process(source);

    expect(result).not.toContain(":root");
    expect(result).toContain("color: #ff0000");
  });

  test("handles multiple var() references in single declaration", () => {
    const source = `
      :root {
        --spacing-x: 10px;
        --spacing-y: 20px;
      }
      .box { margin: var(--spacing-y) var(--spacing-x); }
    `;

    expect(process(source)).toMatchSnapshot();
  });

  test("skips processing when no root vars present", () => {
    const source = `
      .button { 
        color: blue;
        padding: 10px;
      }
      .header {
        font-size: 16px;
      }
    `;

    expect(process(source)).toBe(source.trim());
  });

  test("ignores :root when part of a larger selector", () => {
    const source = `
      :root.dark {
        --primary-color: #000000;
      }
      :root.light {
        --primary-color: #ffffff;
      }
      .button { color: var(--primary-color); }
    `;

    expect(process(source)).toMatchSnapshot();
  });

  test("preserves all rules when complex :root selectors exist", () => {
    const source = `
      :root {
        --primary-color: #ffffff;
      }
      :root.dark {
        --primary-color: #000000;
      }
      .button { color: var(--primary-color); }
    `;

    expect(process(source)).toMatchSnapshot();
  });

  test("preserves rules with multiple complex :root selectors", () => {
    const source = `
      :root {
        --primary-color: #ffffff;
      }
      :root.dark {
        --primary-color: #000000;
      }
      :root[data-theme="custom"] {
        --primary-color: #ff0000;
      }
      .button { color: var(--primary-color); }
    `;

    expect(process(source)).toMatchSnapshot();
  });
});
