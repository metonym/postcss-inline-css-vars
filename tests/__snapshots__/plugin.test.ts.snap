// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`inlineCssVars handles multiple CSS variables 1`] = `
"h1 { color: #ff0000; }
      p { color: #00ff00; }"
`;

exports[`inlineCssVars removes declarations with undefined variables 1`] = `
".button {
        background: #ff0000;
      }"
`;

exports[`inlineCssVars removes declarations with any undefined variable in multi-var declarations 1`] = `
".box {
        padding: 10px 10px;
      }"
`;

exports[`inlineCssVars removes declarations with undefined variables instead of preserving them 1`] = `
"h1 {
        background: #ff0000;
      }"
`;

exports[`inlineCssVars handles multiple var() references in single declaration 1`] = `".box { margin: 20px 10px; }"`;

exports[`inlineCssVars ignores :root when part of a larger selector 1`] = `
":root.dark {
        --primary-color: #000000;
      }
      :root.light {
        --primary-color: #ffffff;
      }
      .button { color: var(--primary-color); }"
`;

exports[`inlineCssVars preserves all rules when complex :root selectors exist 1`] = `
":root {
        --primary-color: #ffffff;
      }
      :root.dark {
        --primary-color: #000000;
      }
      .button { color: var(--primary-color); }"
`;

exports[`inlineCssVars preserves rules with multiple complex :root selectors 1`] = `
":root {
        --primary-color: #ffffff;
      }
      :root.dark {
        --primary-color: #000000;
      }
      :root[data-theme="custom"] {
        --primary-color: #ff0000;
      }
      .button { color: var(--primary-color); }"
`;
